rcl_action/src/rcl_action/action_server.c

```diff
@@ -395,6 +421,66 @@ rcl_action_accept_new_goal(
   return goal_handles[num_goal_handles];
 }
 
+// Implementation only
+static rcl_ret_t
+_recalculate_expire_timer(
+  rcl_timer_t * expire_timer,
+  const int64_t timeout,
+  rcl_action_goal_handle_t ** goal_handles,
+  size_t num_goal_handles,
+  rcl_clock_t * clock)
+{
+  size_t num_inactive_goals = 0u;
+  int64_t minimum_period = 0;
+
+  // Get current time (nanosec)
+  int64_t current_time;
+  rcl_ret_t ret = rcl_clock_get_now(clock, &current_time);
+  if (RCL_RET_OK != ret) {
+    return RCL_RET_ERROR;
+  }
+
+  for (size_t i = 0; i < num_goal_handles; ++i) {
+    rcl_action_goal_handle_t * goal_handle = goal_handles[i];
+    if (!rcl_action_goal_handle_is_active(goal_handle)) {
+      ++num_inactive_goals;
+
+      rcl_action_goal_info_t goal_info;
+      ret = rcl_action_goal_handle_get_info(goal_handle, &goal_info);
```

Comment:
I don't know what's in `goal_info`, but does it need to be `fini` at all?

Reply:
You're right this should have a fini function and it should be used. It's a typedef to a generated `action_msgs` msg which currently doesn't allocate anything on the heap, but there's no guarantee that won't change in the future.

Reply:
There's more to think about here, and I think the solution is to make `rcl_action_goal_info_t` a struct instead of reusing the c struct from action messages, but that's a lot of places to change it so if it's alright I'd like to do this in a follow up PR.